File: ruby/lib/runapi/core/http_client.rb
# frozen_string_literal: true

module RunApi
  module Core
    class HttpClient
      STALE_CONNECTION_ERRORS = [Errno::EPIPE, EOFError, IOError, OpenSSL::SSL::SSLError].freeze

      def initialize(options)
        @options = options
        @pool = ConnectionPool.new(size: 5, timeout: 5) do
          build_connection
        end
      end

      def request(method, path, body: nil, options: nil)
        uri = URI.join(@options.base_url, path)
        req = build_request(method, uri, body, options)
        max_retries = options&.max_retries || @options.max_retries
        retries = 0
        stale_retried = false

        loop do
          response = begin
            @pool.with do |http|
              http.start unless http.started?
              http.request(req)
            end
          rescue *STALE_CONNECTION_ERRORS
            unless stale_retried
              stale_retried = true
              next
            end
            raise NetworkError, "Connection lost"
          rescue ::Net::OpenTimeout, ::Net::ReadTimeout => e
            raise TimeoutError, e.message
          rescue ::SocketError, ::Errno::ECONNREFUSED, ::Errno::ECONNRESET => e
            raise NetworkError, e.message
          end

          if response.is_a?(Net::HTTPNotModified) && options&.allow_not_modified
            return Response.new(body: {"not_modified" => true}, headers: response_headers(response))
          end

          if response.is_a?(Net::HTTPSuccess)
            body = parse_body(response.body)
            return nil if body.nil?
            return body unless body.is_a?(Hash) || body.is_a?(Array)

            return Response.new(body:, headers: response_headers(response))
          end

          error = Error.from_response(response, response.body)

          if retryable?(method, response.code.to_i) && retries < max_retries
            retries += 1
            sleep(retry_delay(retries, error))
            stale_retried = false
            next
          end

          raise error
        end
      ensure
        close_multipart_files(req)
      end

      def close
        @pool.shutdown do |http|
          http.finish if http.started?
        end
      end

      # PUT bytes straight to a pre-authorized upload URL with the exact headers
      # issued for it. Skips the base URL, auth, and retries: the URL is single-use
      # and the body is not safe to replay.
      def upload(url, headers:, body:)
        uri = URI.parse(url)
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = (uri.scheme == "https")
        http.open_timeout = @options.timeout
        http.read_timeout = @options.timeout

        req = Net::HTTP::Put.new(uri.request_uri)
        headers.each { |key, value| req[key.to_s] = value }
        req.body = body

        response = http.start { |connection| connection.request(req) }
        return if response.is_a?(Net::HTTPSuccess)

        raise Error.from_response(response, response.body)
      rescue ::Net::OpenTimeout, ::Net::ReadTimeout => e
        raise TimeoutError, e.message
      rescue ::SocketError, ::Errno::ECONNREFUSED, ::Errno::ECONNRESET => e
        raise NetworkError, e.message
      end

      private

      def build_connection
        uri = URI.parse(@options.base_url)
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = (uri.scheme == "https")
        http.verify_mode = OpenSSL::SSL::VERIFY_PEER
        http.open_timeout = @options.timeout
        http.read_timeout = @options.timeout
        http
      end

      def build_request(method, uri, body, options)
        klass = Net::HTTP.const_get(method.to_s.capitalize)
        req = klass.new(uri.request_uri)

        req["Authorization"] = "Bearer #{@options.api_key}" if @options.api_key
        req["Accept"] = "application/json"
        req["User-Agent"] = Constants::SDK_USER_AGENT

        options&.headers&.each { |k, v| req[k.to_s] = v }

        if body.is_a?(MultipartBody)
          req.set_form(multipart_parts(body), "multipart/form-data")
        elsif body
          req["Content-Type"] = "application/json"
          req.body = JSON.generate(body)
        end
        req
      end

      def multipart_parts(body)
        opened_files = []
        field_parts = body.fields.flat_map do |key, value|
          Array(value).map { |item| [key, item.to_s] }
        end
        file_parts = body.files.map do |key, file|
          options = {filename: file.filename}
          options[:content_type] = file.content_type if file.content_type
          opened_files << File.open(file.path, "rb")
          [key, opened_files.last, options]
        end
        field_parts + file_parts
      rescue
        opened_files.each { |file| file.close unless file.closed? }
        raise
      end

      def close_multipart_files(request)
        body_data = request&.instance_variable_get(:@body_data)
        return unless body_data

        body_data.each do |part|
          file = part[1]
          file.close if file.is_a?(File) && !file.closed?
        end
      end

      def retryable?(method, status)
        Constants::IDEMPOTENT_METHODS.include?(method.to_s.upcase) &&
          Constants::RETRYABLE_STATUS_CODES.include?(status)
      end

      def retry_delay(attempt, error)
        if error.is_a?(RateLimitError) && error.retry_after&.positive?
          return error.retry_after
        end

        base = @options.retry_base_delay * (2**(attempt - 1))
        jitter = rand * base * 0.5
        [base + jitter, @options.retry_max_delay].min
      end

      def parse_body(body)
        return nil if body.nil? || body.empty?

        JSON.parse(body)
      rescue JSON::ParserError
        body
      end

      def response_headers(response)
        headers = {}
        response.each_header { |key, value| headers[key] = value }
        headers
      end
    end
  end
end

File: python/src/runapi/core/http_client.py
"""HTTP transport built on ``httpx.Client`` with retries and multipart support."""

from __future__ import annotations

import json
import random
import time
from typing import Any, Dict, List, Optional, Tuple
from urllib.parse import urlsplit

import httpx

from . import constants
from .errors import NetworkError, RateLimitError, TimeoutError, ValidationError, error_from_response
from .multipart import MultipartBody
from .options import ClientOptions, RequestOptions
from .response import ApiResponse

_NO_BODY = object()


class HttpClient:
    """Synchronous HTTP client.

    ``httpx.Client`` keeps connections alive across requests, so no explicit
    connection pool is needed. Pass ``transport`` to inject an
    ``httpx.MockTransport`` in tests.
    """

    def __init__(self, options: ClientOptions, *, transport: Optional[httpx.BaseTransport] = None) -> None:
        self._options = options
        headers = {
            "Accept": "application/json",
            "User-Agent": constants.SDK_USER_AGENT,
        }
        if options.api_key:
            headers["Authorization"] = f"Bearer {options.api_key}"
        self._client = httpx.Client(
            base_url=options.base_url,
            timeout=options.timeout,
            transport=transport,
            headers=headers,
        )
        # A bare client for direct uploads: the pre-authorized upload URL lives
        # outside the API host and must not receive the API key.
        self._upload_client = httpx.Client(timeout=options.timeout, transport=transport)

    def request(
        self,
        method: str,
        path: str,
        body: Any = None,
        options: Optional[RequestOptions] = None,
    ) -> Any:
        self._validate_request_url(path)
        max_retries = self._options.max_retries
        if options is not None and options.max_retries is not None:
            max_retries = options.max_retries

        headers = {str(key): str(value) for key, value in (options.headers or {}).items()} if options else {}
        timeout = options.timeout if (options and options.timeout is not None) else httpx.USE_CLIENT_DEFAULT

        json_body, data, files, opened = self._build_payload(body)
        method = method.upper()
        retries = 0

        try:
            while True:
                try:
                    response = self._client.request(
                        method,
                        path,
                        json=json_body,
                        data=data,
                        files=files,
                        headers=headers,
                        timeout=timeout,
                    )
                except httpx.TimeoutException as exc:
                    if self._retryable_request(method, headers) and retries < max_retries:
                        retries += 1
                        time.sleep(self._retry_delay(retries, TimeoutError(str(exc))))
                        continue
                    raise TimeoutError(str(exc))
                except httpx.TransportError as exc:
                    if self._retryable_request(method, headers) and retries < max_retries:
                        retries += 1
                        time.sleep(self._retry_delay(retries, NetworkError(str(exc))))
                        continue
                    raise NetworkError(str(exc))

                if response.status_code == 304:
                    return None

                if response.is_success:
                    body = self._parse_body(response.text)
                    if body is None:
                        return None
                    if isinstance(body, (dict, list)):
                        return ApiResponse(body, response.headers, status_code=response.status_code)
                    return body

                error = error_from_response(response)
                if (
                    self._retryable_request(method, headers)
                    and response.status_code in constants.RETRYABLE_STATUS_CODES
                    and retries < max_retries
                ):
                    retries += 1
                    time.sleep(self._retry_delay(retries, error))
                    continue

                raise error
        finally:
            for handle in opened:
                handle.close()

    def upload(self, url: str, headers: Dict[str, str], body: bytes) -> None:
        """PUT bytes straight to a pre-authorized upload URL with the exact headers
        issued for it. No auth, no retries: the URL is single-use and the body is
        not safe to replay."""
        try:
            response = self._upload_client.put(url, content=body, headers=headers)
        except httpx.TimeoutException as exc:
            raise TimeoutError(str(exc))
        except httpx.TransportError as exc:
            raise NetworkError(str(exc))

        if not response.is_success:
            raise error_from_response(response)

    def request_bytes(
        self,
        method: str,
        path: str,
        options: Optional[RequestOptions] = None,
    ) -> bytes:
        """Return a successful response body without text decoding."""
        self._validate_request_url(path)
        max_retries = self._options.max_retries
        if options is not None and options.max_retries is not None:
            max_retries = options.max_retries
        headers = {str(key): str(value) for key, value in (options.headers or {}).items()} if options else {}
        timeout = options.timeout if (options and options.timeout is not None) else httpx.USE_CLIENT_DEFAULT
        method = method.upper()
        retries = 0

        while True:
            try:
                response = self._client.request(method, path, headers=headers, timeout=timeout)
            except httpx.TimeoutException as exc:
                raise TimeoutError(str(exc))
            except httpx.TransportError as exc:
                raise NetworkError(str(exc))
            if response.is_success:
                return response.content
            error = error_from_response(response)
            if (
                self._retryable_request(method, headers)
                and response.status_code in constants.RETRYABLE_STATUS_CODES
                and retries < max_retries
            ):
                retries += 1
                time.sleep(self._retry_delay(retries, error))
                continue
            raise error

    def close(self) -> None:
        self._client.close()
        self._upload_client.close()

    def _build_payload(
        self, body: Any
    ) -> Tuple[Any, Optional[Dict[str, Any]], Optional[Dict[str, Any]], List[Any]]:
        if isinstance(body, MultipartBody):
            files = []
            for key, value in body.fields.items():
                values = value if isinstance(value, (list, tuple)) else [value]
                files.extend((key, (None, str(item))) for item in values)
            opened: List[Any] = []
            for key, part in body.files.items():
                handle = open(part.path, "rb")
                opened.append(handle)
                filename = part.filename
                if part.content_type:
                    files.append((key, (filename, handle, part.content_type)))
                else:
                    files.append((key, (filename, handle)))
            return None, None, files, opened
        if body is not None:
            return body, None, None, []
        return None, None, None, []

    def _retryable_request(self, method: str, headers: Dict[str, str]) -> bool:
        return method in constants.IDEMPOTENT_METHODS or (
            method == "POST"
            and any(name.lower() == "idempotency-key" and str(value).strip() for name, value in headers.items())
        )

    def _validate_request_url(self, path: str) -> None:
        requested = urlsplit(path)
        if not requested.scheme and not requested.netloc:
            return
        configured = urlsplit(str(self._options.base_url))
        if self._origin(requested) != self._origin(configured):
            raise ValidationError("Request URL must use the configured RunAPI origin")

    @staticmethod
    def _origin(url: Any) -> Tuple[str, str, Optional[int]]:
        scheme = url.scheme.lower()
        port = url.port or ({"http": 80, "https": 443}.get(scheme))
        return scheme, (url.hostname or "").lower(), port

    def _retry_delay(self, attempt: int, error: Any) -> float:
        if isinstance(error, RateLimitError) and error.retry_after and error.retry_after > 0:
            return error.retry_after
        base = self._options.retry_base_delay * (2 ** (attempt - 1))
        jitter = random.random() * base * 0.5
        return min(base + jitter, self._options.retry_max_delay)

    @staticmethod
    def _parse_body(body: Optional[str]) -> Any:
        if not body:
            return None
        try:
            return json.loads(body)
        except ValueError:
            return body
